pkg/rfc8693/claims.go

```diff
@@ -0,0 +1,92 @@
+package rfc8693
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+
+	"github.com/google/cel-go/cel"
+	"github.com/ory/fosite/token/jwt"
+)
+
+const (
+	celVariableClaims    = "claims"
+	celVariableSubSHA256 = "subSHA256"
+)
+
+func parseCEL(input string) (cel.Program, error) {
+	env, err := cel.NewEnv(
+		cel.Variable(celVariableClaims, cel.MapType(cel.StringType, cel.DynType)),
+		cel.Variable(celVariableSubSHA256, cel.StringType),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ast, issues := env.Compile(input)
+	if err := issues.Err(); err != nil {
+		return nil, err
+	}
+
+	prog, err := env.Program(ast)
+	if err != nil {
+		return nil, err
+	}
+
+	return prog, nil
+}
+
+// ClaimMappingStrategy represents a mapping from external identity claims to DMV claims.
+type ClaimMappingStrategy struct {
+	mappings map[string]cel.Program
+}
+
+// NewClaimMappingStrategy creates a ClaimMappingStrategy given a mapping of desired DMV claims to CEL expressions.
+func NewClaimMappingStrategy(mappingExprs map[string]string) (ClaimMappingStrategy, error) {
+	mappings := make(map[string]cel.Program, len(mappingExprs))
+
+	for k, e := range mappingExprs {
+		prog, err := parseCEL(e)
+		if err != nil {
+			return ClaimMappingStrategy{}, err
+		}
+
+		mappings[k] = prog
+	}
+
+	out := ClaimMappingStrategy{
+		mappings: mappings,
+	}
+
+	return out, nil
+}
+
+// MapClaims consumes a set of JWT claims and produces a new set of mapped claims.
+func (m ClaimMappingStrategy) MapClaims(claims jwt.JWTClaims) (jwt.JWTClaims, error) {
```

Comment:
Was there a specific reason this returns a value? Every time the chain of callers pass around the `jwt.JWTClaims`, the value will be copied. Would it make more sense to return a pointer instead to avoid this?

Reply:
No specific reason. This should be fixed now to use `*jwt.JWTClaims` where needed and `jwt.JWTClaimsContainer` where appropriate.